### PR TITLE
Add relevant icons to all types of status messages and modals #7709

### LINKED ISSUE #7709: Add relevant icons to all types of status messages and modals
While trying to identify the cause for #7708, I found that in the discussion for PR #7067, @damithc had recommended that in addition to adding a `tick` before each SUCCESS status message, we also **add relevant icons to all types of status messages**. Since I couldn't find an existing issue to this effect, I created this. 

Note: adding icons to status messages will have to wait until #7708 is solved to preserve the sanctity of all our modals.

### PR DESCRIPTION
**Pull request by HirdayGupta**
Fixes #7709

Code in this PR will make sense after #7774 is merged.

**Outline of Solution:**

- Added relevant icons to all status messages using the new icon classes. The correct `icon` classes are being added due to #7774. I am now making use of them to add the correct icons.
- I wonder if there's a way to reuse the repeated code for all the status icons? All of these statuses need the same properties except `content`. Should I consider adding an `icon` class which adds the basic properties and then the specific icon classes set the content?

### CODE REVIEW

File: src/main/webapp/stylesheets/teammatesCommon.css [hunk 1]
```diff
@@ -1193,3 +1193,27 @@ Glyphicon for success status messages
     padding-right: 2px;
     content: '\e084';
 }
+
+.icon-warning:before {
```
Comment (whipermr5) [inline src/main/webapp/stylesheets/teammatesCommon.css]: You can stack the selectors:
```css
.icon-warning:before,
.icon-danger:before,
.icon-info:before {
    ... // common settings
}
```

**Comment by HirdayGupta** [comment]
Dropping a note that I have rebased my branch and force pushed. Also, @whipermr5, thanks for the suggestion! I have stacked the selectors in 55bc5a5.

File: src/main/webapp/stylesheets/teammatesCommon.css [hunk 2]
```diff
@@ -1187,11 +1187,27 @@ Prevent Long URL From Breaking Out of Container
 /*
 Glyphicon for success status messages
 */
-
```
Comment (whipermr5) [inline src/main/webapp/stylesheets/teammatesCommon.css]: Restore this blank line to separate the heading from the content.

File: src/main/webapp/stylesheets/teammatesCommon.css [hunk 3]
```diff
@@ -1187,11 +1187,27 @@ Prevent Long URL From Breaking Out of Container
 /*
 Glyphicon for success status messages
```
Comment (whipermr5) [inline src/main/webapp/stylesheets/teammatesCommon.css]: Can follow the big block heading style used above (refer to the `Responsiveness` section).

File: src/main/webapp/stylesheets/teammatesCommon.css [hunk 4]
```diff
@@ -1187,11 +1187,27 @@ Prevent Long URL From Breaking Out of Container
...
+.icon-warning:before,
+.icon-danger:before,
+.icon-info:before {
     font-family: 'Glyphicons Halflings';
     display: inline-block;
     padding-right: .5em;
-    content: '\e084';
     vertical-align: top;
 }
+.icon-success:before {
```
Comment (whipermr5) [inline src/main/webapp/stylesheets/teammatesCommon.css]: Add a blank line before this.

**Comment by HirdayGupta** [comment]
Now that #7774 is merged, here are screenshots of the icons:

### Success Icons:

**Modal:**
screen shot 2017-07-25 at 2 16 04 pm

**Status:**
screen shot 2017-07-26 at 1 05 33 pm

### Warning Icons:

**Modal:**
screen shot 2017-07-26 at 12 58 07 pm

**Status:**
screen shot 2017-07-26 at 1 00 47 pm

### Danger Icons:

**Modal:**
screen shot 2017-07-26 at 12 58 43 pm

**Status:**
screen shot 2017-07-26 at 1 01 31 pm


### Info Icons:

**Modal**
screen shot 2017-07-26 at 1 00 16 pm

(Could not find an info status message 😅 )

**Comment by HirdayGupta** [comment]
@whipermr5 Ready for review :)

File: src/main/webapp/stylesheets/teammatesCommon.css [hunk 5]
```diff
@@ -1184,14 +1184,32 @@ Prevent Long URL From Breaking Out of Container
...
+/*//////////////////////////////////////////////////////////
+////////////////////// Status Icons //////////////////////
+//////////////////////////////////////////////////////////*/
 
-.icon-success:before {
+.icon-success:before,
+.icon-warning:before,
+.icon-danger:before,
+.icon-info:before {
     font-family: 'Glyphicons Halflings';
```
Comment (LiHaoTan) [inline src/main/webapp/stylesheets/teammatesCommon.css]: Can you add `/*noinspection CssNoGenericFontName*/` here to suppress warnings (in IntelliJ)?

Comment (HirdayGupta) [inline src/main/webapp/stylesheets/teammatesCommon.css]: Sure. That would be after `line 1195`?

Comment (LiHaoTan) [inline src/main/webapp/stylesheets/teammatesCommon.css]: Before

**Comment by HirdayGupta** [comment]
@whipermr5 Review reminder :)

File: src/main/webapp/stylesheets/teammatesCommon.css [hunk 6]
```diff
@@ -1184,14 +1184,33 @@ Prevent Long URL From Breaking Out of Container
...
+/*//////////////////////////////////////////////////////////
+////////////////////// Status Icons //////////////////////
+//////////////////////////////////////////////////////////*/
 
-.icon-success:before {
+.icon-success:before,
+.icon-warning:before,
+.icon-danger:before,
+.icon-info:before {
+    /*noinspection CssNoGenericFontName*/
```
Comment (whipermr5) [inline src/main/webapp/stylesheets/teammatesCommon.css]: @LiHaoTan I'm not sure we should add a suppression just to turn off a warning that is very specific to IntelliJ. Doing this implies that we regard the default inspection settings of a single IDE as authoritative, and always to be followed whenever we write code. This is outside the spirit that our project is IDE-agnostic. Only general warnings that apply to most IDEs should be considered.

Comment (LiHaoTan) [inline src/main/webapp/stylesheets/teammatesCommon.css]: I don't think I agree on this aspect if we are supporting the IDEs. There's a reason why IntelliJ (and other IDEs) made their inspection suppressions as comments, so they become just comments outside of the IDE.

Being IDE-agnostic does not mean we should throw away good features that an IDE offer unless they specifically conflict with the workflow. IntelliJ has taken the efforts to specifically make suppressing IntelliJ specific inspections comments so it does not interfere with the workflow.

As for whether we treat the default settings as authoritative, that's up to the project? We should turn those inspections off completely if we don't agree with it, not ignore them and leave the warnings hanging around.

As for this particular instance, I think IntelliJ's defaults are sane here. There is a general consensus that it is a good thing to add generic font families and it is also recommended by various sources and specifications:
>  Authors are encouraged to append a generic font family as a last alternative for improved robustness.

https://drafts.csswg.org/css-fonts-4/#font-family-prop
https://www.w3.org/TR/css-fonts-3/#font-family-prop

> Note: You should always include at least one generic family name in a font-family list, since there's no guarantee that any given font is available. This lets the browser select an acceptable fallback font when necessary.
https://developer.mozilla.org/en/docs/Web/CSS/font-family


So as this is a special case where it does not make sense that an icon font have generic fallback, I think suppressing the warning is the way to go.

Comment (LiHaoTan) [inline src/main/webapp/stylesheets/teammatesCommon.css]: I don't disagree with the fact that if IDEs can be made optional it can be nice though.

Though saying that an IDE's usage should never be a requirement is too far of a stretch (some projects may simply want consistency throughout and depend on the IDE tools to achieve some things, and they can't be achieved outside of the IDE, or at least easily).

Comment (whipermr5) [inline src/main/webapp/stylesheets/teammatesCommon.css]: > There's a reason why IntelliJ (and other IDEs) made their inspection suppressions as comments, so they become just comments outside of the IDE.

They aren't really the same as normal comments. For one thing, they make code less readable rather than more readable.

Agree with your points about no throwing away good features; however, my main point is that to introduce such specific suppressions, we'd need to come to a decision which IDE default inspections we'd like to adopt, then consistently follow them.

> As for whether we treat the default settings as authoritative, that's up to the project?

Exactly - the problem is, our project doesn't have any such recommendation right now. These things are considered to be user-specific.

> We should turn those inspections off completely if we don't agree with it, not ignore them and leave the warnings hanging around.

People who use Eclipse or don't use an IDE won't be bothered by them (instead, they'd be bothered by unnecessary comments that are suppressions, and wouldn't have any option to hide them). People who use IntelliJ and are bothered by warnings have the options of turning off the specific inspections that generate those warnings, or living with the warnings.

To sum up, these things are all part of the specific local setup that we don't dictate. I, too, am an IntelliJ fan. However, adding stuff to the codebase (comments or otherwise) for the express purpose of conforming to a certain IDE is what I'm against. I'll only have it if we specifically decide to elevate one IDE's authority over the others, then consistently stick to it throughout the codebase. Doing so would indeed bring about benefits, though at the cost of additional burden to document and enforce these recommendations (e.g. all reviewers would have to open the code in IntelliJ to check that it didn't trigger another warning, then tell the developer to add in a specific suppression comment before approving the PR).

Comment (HirdayGupta) [inline src/main/webapp/stylesheets/teammatesCommon.css]: I agree with @whipermr5. Since we don't recommend only a single IDE, our codebase should only include lines that are to do with our Application and nothing else. Plus, I don't know about IntelliJ, but at least in Eclipse, as of now, there are a lot of warnings too. If we were to identify and suppress each of them our code would not only become messy but also unreadable. This argument is compounded by the fact that IDE warnings are generally easy to ignore. I don't think this like should get special treatment.

Comment (LiHaoTan) [inline src/main/webapp/stylesheets/teammatesCommon.css]: Well, for one thing, it is going to be difficult to start choosing which sets of defaults we want to choose in one go. I rather we do it slowly, for example, if in this case, we decide that this inspection is to be turned off, we turn it off in this PR.

As for making the code unreadable, that is only if the comment is useless, which in this case it isn't. It points out something is not wanted, which in some cases the comment may not be clear on what it is doing we can explain it.

Now, why not no comments altogether so users do not have to be distracted at all. However, in this case, no matter whether a person is using an IDE or not it isn't a distraction because we are not following something commonly recommended in CSS, and we should be pointing it out.

Comment (LiHaoTan) [inline src/main/webapp/stylesheets/teammatesCommon.css]: As for the project having tons of warnings in Eclipse does not translate to saying we should have warnings everywhere (in fact IntelliJ IDEA gives lots of warnings for the project as well, and I would say they are mostly valid (if not all), the last time I scanned through the entire project).

Comment (whipermr5) [inline src/main/webapp/stylesheets/teammatesCommon.css]: @wkurniawan07 @chowyb @leeyimin @thenaesh What do y'all think? :)

Comment (HirdayGupta) [inline src/main/webapp/stylesheets/teammatesCommon.css]: @wkurniawan07 @chowyb @leeyimin @thenaesh Awaiting your inputs :)

Comment (whipermr5) [inline src/main/webapp/stylesheets/teammatesCommon.css]: @LiHaoTan Perhaps you can open a new issue to discuss this so that this PR can continue. For now, I feel `/*noinspection CssNoGenericFontName*/` should be removed. The only inspections we currently enforce are those performed by our current static analysis tools. If you want to suggest enforcing other inspections, it can be put up for discussion in the issue tracker.

Comment (LiHaoTan) [inline src/main/webapp/stylesheets/teammatesCommon.css]: Yea I agree there are some issues surrounding adding `/*noinspection CssNoGenericFontName*/` (I will point them out in the new issue accordingly).

Not IntelliJ specific however I do think there should be a comment like
`/* icon font, generic fallback not relevant */`. However, it comes off as strange because without a lint warning there is no context unless the developer knows why (we can't assume that), yet not including it means that we don't care about generic fallbacks at all (which I think is a bigger mistake).

Personally, then I think the best compromise I can think of is to include the IntelliJ specific inspection suppression.

If there is a better compromise that does not involve the IntelliJ specific inspection, I would definitely prefer it.

Comment (whipermr5) [inline src/main/webapp/stylesheets/teammatesCommon.css]: I guess I'm okay with this:
```css
.icon-success:before,
.icon-warning:before,
.icon-danger:before,
.icon-info:before {
    /*Icon font has no generic fallback*/
    /*noinspection CssNoGenericFontName*/
    font-family: 'Glyphicons Halflings';
    display: inline-block;
    padding-right: .5em;
    vertical-align: top;
}
```
It'd look nicer with padding using spaces though, which still works even though it differs from what IntelliJ inserts when you select `Suppress for statement` from the IntelliJ menu:
```css
.icon-success:before,
.icon-warning:before,
.icon-danger:before,
.icon-info:before {
    /* Icon font has no generic fallback */
    /* noinspection CssNoGenericFontName */
    font-family: 'Glyphicons Halflings';
    display: inline-block;
    padding-right: .5em;
    vertical-align: top;
}
```
Mixing both styles looks weird though:
```css
.icon-success:before,
.icon-warning:before,
.icon-danger:before,
.icon-info:before {
    /* Icon font has no generic fallback */
    /*noinspection CssNoGenericFontName*/
    font-family: 'Glyphicons Halflings';
    display: inline-block;
    padding-right: .5em;
    vertical-align: top;
}
```
@LiHaoTan are you okay with the middle option? Or do you think that's unreliable as adding spaces might cause the suppression to break if IntelliJ changes their implementation in a future version?

Comment (LiHaoTan) [inline src/main/webapp/stylesheets/teammatesCommon.css]: Yea I think the middle option looks better too.

**Comment by HirdayGupta** [comment]
@whipermr5 Ready for review :)

File: src/main/webapp/stylesheets/teammatesCommon.css (hunk 6 above)
Comment (wkurniawan07) [inline src/main/webapp/stylesheets/teammatesCommon.css]: Looks like I'm late to the party, but my take here is a vote against the inclusion of such line.

> To sum up, these things are all part of the specific local setup that we don't dictate. I, too, am an IntelliJ fan. However, adding stuff to the codebase (comments or otherwise) for the express purpose of conforming to a certain IDE is what I'm against. I'll only have it if we specifically decide to recommend one IDE and its default inspection profile over the others, then consistently stick to it throughout the codebase.

This line by @whipermr5 pretty much sums up my stand. It is difficult to enforce even _one_ IDE profile; you'd have noticed by now that the defaults of any IDE lack many good features of (at least Java) languages. My own Eclipse inspection profile has many more additional rules on top of the default ones, and I'm not sure how many Eclipse violations are detected out there, but I got 245 right now (many of which I'm itching to fix). I'm not saying that you should throw away good IDE features, but unless they are general enough (e.g other IDEs will also complain anyway; CheckStyle/... will complain anyway), I don't think we have a ground to include any IDE-specific annotations around. It is like saying "I [a person] don't understand this code, let's add a comment here" while the majority of people out there have no problem understanding it.

Comment (LiHaoTan) [inline src/main/webapp/stylesheets/teammatesCommon.css]: > This line by @whipermr5 pretty much sums up my stand. It is difficult to enforce even one IDE profile; you'd have noticed by now that the defaults of any IDE lack many good features of (at least Java) languages.

Well, I definitely do not think adding IDE specific comments is a good idea as well, but at the moment the most reasonable compromise I can think of is what I have discussed.


If there is an equivalent check that can be integrated into CI, we can very well turn that inspection off in IntelliJ (but it does not seem any popular tool does this).

>  I'm not saying that you should throw away good IDE features, but unless they are general enough (e.g other IDEs will also complain anyway; CheckStyle/... will complain anyway)

I think I have made arguments for that it is general enough unless is there some reason you don't think so?

> My own Eclipse inspection profile has many more additional rules on top of the default ones, and I'm not sure how many Eclipse violations are detected out there, but I got 245 right now (many of which I'm itching to fix).

Violations in Eclipse that are general enough should be reflected in IntelliJ as well, so we could always fix them (and if they aren't, we should turn those checks off). For example, using the diamond constructor for Java is definitely a good idea, but `CheckStyle` doesn't have a way to check for that.

That should be similar in this case.

> I don't think we have a ground to include any IDE-specific annotations around. It is like saying "I [a person] don't understand this code, let's add a comment here" while the majority of people out there have no problem understanding it.

I agree there certainly needs to be some kind of assumed knowledge so comments don't end up mentioning for example why inheritance is used but at least I think this is not the case here.

Do the majority of people understand that fonts should have generic fallbacks, and thus icon fonts are special cases? Assuming this knowledge does not seem ideal, but while that can be arguable, a linter would help us with avoid this kind of problem altogether.

So logically if we want to assume the user have that knowledge, we simply mention that in our style guide. And if we violate our style guide, we will logically need to mention why we choose to ignore our style guide (so it's like how people also serve the linting role when we don't follow our style guide).

But this causes the problem that the warnings still persist in IntelliJ then, should we then turn the inspection off and let people manually check, but that does not sound like a very good idea either.

So given that IntelliJ is the only one who gives such a warning right now, and that warning is general enough that's why I think the most reasonable compromise would be https://github.com/TEAMMATES/teammates/pull/7844#discussion_r130527179

--- 

Basically, I am not saying that this is a good solution, but it is the most reasonable compromise I can think of (otherwise there won't be a need for discussion anyway). I am open to a better alternative.

Ultimately, I know of a possible long term solution, but just not sure whether we can actually achieve it. We can use IntelliJ inspections as a supplement to the current static analysis tools we have, by integrating it into CI as well. Doing it through `TeamCity` would be reasonably, but the problem is that our project doesn't seem to qualify for the open source hosted version, and I'm not sure if self-hosting would be a good idea.

Installing IntelliJ and then running the inspections on Travis seems possible, but I'm not sure if we actually need the Ultimate version to have all the inspections.

**Comment by whipermr5** [comment]
@damithc Reminder just in case you missed this :)

**Comment by HirdayGupta** [comment]
@damithc Reminder :)
